File: filtering.py
import sys

def process_wordlist(input_file, output_file):
    print("Memulai program\n")    

    # Menggunakan set untuk menyimpan baris unik
    unique_lines = set()

    print("Memperkirakan jumlah proses yang akan dikerjakan")

    # Menghitung jumlah total baris untuk progress bar
    with open(input_file, 'r', encoding='latin1') as infile:
        total_lines = sum(1 for line in infile)

    # Membaca file dan menyaring baris yang memenuhi kriteria
    with open(input_file, 'r', encoding='latin1') as infile:
        for i, line in enumerate(infile, 1):
            line = line.strip()  # Menghapus whitespace di awal/akhir
            if len(line) >= 8:  # Memeriksa panjang baris
                unique_lines.add(line)

            # Update progress bar
            progress = int(50 * i / total_lines)
            sys.stdout.write(f'\rProcessing: [{"#" * progress}{"." * (50 - progress)}] {i}/{total_lines} lines')
            sys.stdout.flush()

    # Menulis hasil ke file output
    with open(output_file, 'w', encoding='latin1') as outfile:
        for line in sorted(unique_lines):  # Menulis dalam urutan terurut jika diinginkan
            outfile.write(line + '\n')

    print("\nProcessing complete.")

# Memasukan nama wordlist file
file_name = str(input('Nama file [.txt]: '))

# Ganti 'wordlist.txt' dengan nama file input Anda
# Ganti 'cleaned_wordlist.txt' dengan nama file output yang diinginkan
process_wordlist(file_name, 'output.txt')

File: filtering.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>
#include <string>

void process_wordlist(const std::string& input_file, const std::string& output_file) {
    std::ifstream infile(input_file);
    if (!infile.is_open()) {
        std::cerr << "Error opening input file." << std::endl;
        return;
    }

    std::ofstream outfile(output_file);
    if (!outfile.is_open()) {
        std::cerr << "Error opening output file." << std::endl;
        return;
    }

    // Calculate total file size for progress indication
    infile.seekg(0, std::ios::end);
    size_t total_size = infile.tellg();
    infile.seekg(0, std::ios::beg);

    std::string line;
    size_t processed_size = 0;
    std::unordered_set<std::string> unique_lines;

    while (getline(infile, line)) {
        processed_size += line.size() + 1; // +1 for newline character
        if (line.size() >= 8) {
            unique_lines.insert(line);
        }

        // Update progress bar
        int progress = static_cast<int>(50 * processed_size / total_size);
        std::cout << "\rProcessing: [";
        for (int i = 0; i < progress; ++i) std::cout << "#";
        for (int i = progress; i < 50; ++i) std::cout << ".";
        std::cout << "] " << processed_size << "/" << total_size << " bytes";
        std::cout.flush();
    }
    infile.close();

    // Write unique lines directly to the output file
    for (const auto& str : unique_lines) {
        outfile << str << '\n';
    }
    outfile.close();

    std::cout << "\nProcessing complete." << std::endl;
}

int main() {
    std::string file_name;
    std::cout << "Nama file [.txt]: ";
    std::cin >> file_name;

    process_wordlist(file_name, "output.txt");

    return 0;
}
